File: ezy_func_model.py


import numpy as np
import typing as tp
from collections.abc import Sequence
from kron import get_gauss_points

__all__ = ["integrate"]
coeffs = [ 0.          ,-0.9148286559, 0.0561389246,-0.0999854514,-0.0515047232,
  0.0551857439,-0.0036181641,-0.0033739435,-0.001450217 ,-0.0021050108]
intercept  = -0.6157187001159432
limits = [(1e-05, 0.1), (1.0, 100000.0)]


        
def eval_poly(vars: Sequence[float]) -> float:
    return 1 * coeffs[0] + vars[0] * coeffs[1] + vars[1] * coeffs[2] + vars[0] ** 2 * coeffs[3] + vars[0] * vars[1] * coeffs[4] + vars[1] ** 2 * coeffs[5] + vars[0] ** 3 * coeffs[6] + vars[0] ** 2 * vars[1] * coeffs[7] + vars[0] * vars[1] ** 2 * coeffs[8] + vars[1] ** 3 * coeffs[9] + intercept
    

lb: int = 1
ub: int = 15

def integrate(f: tp.Callable[[float, tp.Tuple[float, ...]], float], a: float, b: float, params: tp.Tuple[float, ...] = ()) -> float:

    expo = int(eval_poly([np.log2(max(limit[0], min(limit[1],param))) for limit, param in zip(limits,params)]) + 1) 
    n = int(pow(2, max(lb, min(ub, expo))))

    xg, wg = get_gauss_points(n)

    if len(params) > 1:
        params = params[1:]
    y = (b-a)*(xg+1)/2 + a
    return (b-a)/2 * np.sum(wg*f(y, *params))
    

File: ezy_func_model.c

#include <stdio.h>
#include <stdlib.h>
#include <math.h>
#include "kron.h"

const double coeffs[] = {0.0, -0.9148286558753346, 0.05613892455508609, -0.09998545138804422, -0.051504723211393426, 0.05518574393180811, -0.0036181641266642695, -0.003373943473984037, -0.0014502169867654235, -0.0021050108055385985};
const double intercept = -0.6157187001159432;
const double limits[][2] = {{1e-05, 0.1}, {1.0, 100000.0}};


        
double eval_poly(double var0, double var1){
    return 1 * coeffs[0] + var0 * coeffs[1] + var1 * coeffs[2] + (var0 * var0) * coeffs[3] + var0 * var1 * coeffs[4] + (var1 * var1) * coeffs[5] + (var0 * var0 * var0) * coeffs[6] + (var0 * var0) * var1 * coeffs[7] + var0 * (var1 * var1) * coeffs[8] + (var1 * var1 * var1) * coeffs[9] + intercept;
}
    

const int lb = 1;
const int ub = 15;
typedef double (*Integrand)(double x, double b);

double integrate(Integrand f, double a, double b, double rtol, double b){


    // Determine the number of points for the Gauss quadrature
    int expo = (int)(eval_poly(log2(max(limits[0][0],min(limits[0][1], rtol))), log2(max(limits[1][0],min(limits[1][1], b)))) + 1);
    int n = (int)(pow(2, max(lb, min(ub, expo))));
    
    double *xg, *wg;
    get_gauss_points(n, &xg, &wg);
    
    // Perform the integration
    double sum = 0;
    for (int i = 0; i < n; i++){
        sum += f(a + (b - a) * 0.5 * (xg[i] + 1), b) * wg[i];
    }
    sum *= (b - a) * 0.5;
    return sum;
    
}